decisions/0003-create-a-server-runtime-agnostic-package.md

```diff
@@ -0,0 +1,71 @@
+# Create a server-runtime-agnostic package
+
+Date: 2022-05-19
+
+Status: proposed
+
+## Context
+
+We have three server runtime packages:
+- `@remix-run/cloudflare`
+- `@remix-run/deno`
+- `@remix-run/node`
+
+Each of these re-exports a bunch of values and functions from `@remix-run/server-runtime`.
+
+Each also uses server runtime authoring utilities from `@remix-run/server-runtime` to implement things like `createCookie` in a runtime-specific way.
+
+Currently, we recommend that users import all server-related values and functions from their server runtime package, for example:
+
+```tsx
+import { json } from "@remix-run/node";
+```
+
+Most of these imports are runtime-agnostic, like `json`.
+
+Importing runtime-agnostic values and functions through a specific server runtime package makes user code in `app/` **less portable**; users would have to change those imports if they change their server runtime.
+
+Additionally, maintaining parity for re-exports across all server runtime packages is tedious and error-prone (for example, [having to patch in more re-exports in `@remix-run/deno`](https://github.com/remix-run/remix/pull/3215)).
+
+## Decision
+
+Separate `@remix-run/server-runtime` into two packages:
+- `@remix-run/server-runtime-toolkit`: utilities for authoring server runtime packages (e.g. `createCookieFactory`)
+- `@remix-run/server`: provides runtime-agnostic values and function (e.g. `json`)
+
+Specific package names TBD (`@remix-run/server` might be confusing with `@remix-run/serve` already existing and doing something categorically different).
+
+`@remix-run/server-runtime-toolkit` should not be considered public API, while `@remix-run/server` should be.
+
+## Consequences
+
+### How to import from Remix server packages
+
+Users would import runtime-agnostic values and functions from `@remix-run/server`:
```

Comment:
I don't know if this is desired usage tbh.

Benefits I see:
- We're more transparant about everything
- We need no tests to see that all functions/types/values are exported by all 3 runtimes.

However, the disadvantages are:
- People need to care about more than just the runtime they're using
- It will become harder for us to have runtime specific implementations of certain things
  We currently have runtime specific implementations for `createCookie`, `createCookieSessionStorage`, `createMemorySessionStorage`, ` createSessionStorage` & `createRequestHandler` (only specific in `deno` runtime).
  We would either loose the disability to have this because we're putting all common functions/types/values in 1 package
  _OR_ we would cause confusion because people won't know what would be in the generic package & what would be in the runtime package.

Especially the last disadvantage is a huge one imo.

Reply:
To me it sounds like your points of "People need to care..." and "we would cause confusion because..." are the same thing, but correct me if I'm wrong.

I don't think it would be confusing nor limiting. We can simply follow the rule "If any function has at least 1 runtime-specific implementation, its part of the server runtime interface". For example, `createRequestHandler`.

But the _all_ types from `server-runtime` and the vast majority of values/functions are _never_ runtime-specific.

For users, that means if you are trying to use:
- `createCookie`
- `createCookieSessionStorage`
- `createMemorySessionStorage`
- `createSessionStorage`
- `createRequestHandler`
then grab those from your server runtime package.

For everything else, grab it from `@remix-run/server`.

Reply:
What about the fact that we start as an agnostic package (which was the case with `createRequestHandler` before we had support for `deno`) and later on decide it needs a platform specific implementation?

Reply:
In hindsight, I think `createRequestHandler` is something that should have always been considered platform specific. With the experience we have over the past several months, I think it would be possible to determine if a new API is a common utility function, or something that could be platform specific.

I just don't think it would be a good idea to ditch this proposal because of past "mistakes".

Reply:
> What about the fact that we start as an agnostic package (which was the case with `createRequestHandler` before we had support for `deno`) and later on decide it needs a platform specific implementation?

I think those cases will be rare considering there aren't more than a handful of JS runtimes, but if we find out something needs to be runtime-specific, we can put it into each runtime and provide a codemod for it. Should be easy now that we have existing codemods for changing where things get imported from 😄 But yea, I expect that to be quite rare if it happens at all.